### ISSUE TITLE
Reproducing results of the paper with iMAML

### ISSUE THREAD

**Issue by zaccharieramzi** [prolearner/hypergrad#10]
Hi,

Thanks so much for your work, it's very nice and useful.
I am at the moment struggling with trying to reproduce the results of the iMAML paper anyway I can (I tried with the original implementation, [see issue here](https://github.com/aravindr93/imaml_dev/issues/3) and I am developing my own atm).
I tried with [your implementation](https://github.com/prolearner/hypertorch/blob/master/examples/iMAML.py) and got the following results after quite some outer steps:

```
MT k=37500 (1.726s F: 0.877s, B: 0.847s) Val Loss: 1.80e-02, Val Acc: 99.50.
Test loss 1.04e-01 +- 1.51e-01: Test acc: 96.71 +- 4.27e+00 (mean +- std over 1000 tasks).
MT k=37510 (1.758s F: 0.875s, B: 0.880s) Val Loss: 3.04e-02, Val Acc: 99.33.
MT k=37520 (1.767s F: 0.877s, B: 0.888s) Val Loss: 2.14e-02, Val Acc: 99.42.
MT k=37530 (1.791s F: 0.880s, B: 0.910s) Val Loss: 4.65e-02, Val Acc: 98.33.
MT k=37540 (1.814s F: 0.882s, B: 0.929s) Val Loss: 3.73e-02, Val Acc: 98.50.
MT k=37550 (1.715s F: 0.877s, B: 0.835s) Val Loss: 2.39e-02, Val Acc: 99.17.
MT k=37560 (1.709s F: 0.873s, B: 0.834s) Val Loss: 2.13e-02, Val Acc: 99.08.
MT k=37570 (1.724s F: 0.877s, B: 0.845s) Val Loss: 4.14e-02, Val Acc: 98.17.
MT k=37580 (1.742s F: 0.881s, B: 0.858s) Val Loss: 6.90e-02, Val Acc: 98.08.
MT k=37590 (1.697s F: 0.879s, B: 0.815s) Val Loss: 2.95e-02, Val Acc: 98.67.
MT k=37600 (1.742s F: 0.874s, B: 0.865s) Val Loss: 2.82e-02, Val Acc: 99.08.
MT k=37610 (1.784s F: 0.877s, B: 0.904s) Val Loss: 4.80e-02, Val Acc: 98.50.
MT k=37620 (1.695s F: 0.877s, B: 0.816s) Val Loss: 5.39e-02, Val Acc: 98.50.
MT k=37630 (1.760s F: 0.873s, B: 0.885s) Val Loss: 6.28e-02, Val Acc: 97.92.
MT k=37640 (1.737s F: 0.880s, B: 0.855s) Val Loss: 2.16e-02, Val Acc: 99.33.
MT k=37650 (1.759s F: 0.880s, B: 0.877s) Val Loss: 2.55e-02, Val Acc: 99.08.
MT k=37660 (1.707s F: 0.877s, B: 0.828s) Val Loss: 2.95e-02, Val Acc: 99.08.
MT k=37670 (1.808s F: 0.873s, B: 0.933s) Val Loss: 3.21e-02, Val Acc: 98.92.
MT k=37680 (1.821s F: 0.875s, B: 0.944s) Val Loss: 6.20e-02, Val Acc: 97.67.
MT k=37690 (1.737s F: 0.877s, B: 0.858s) Val Loss: 4.85e-02, Val Acc: 98.58.
```

As you can see the test accuracy is quite below that reported in the paper (96.7% vs 99.1%) and the std is very high (4.27 vs 0.35).

I was wondering if you had managed to reproduce the results of the original paper.

I also had a few questions regarding your implementation choices:
- you use a batch size of 16. I saw that [the original implementation uses 32](https://github.com/aravindr93/imaml_dev/tree/master/examples). How did you chose that number?
- Could you explain what you mean [here](https://github.com/prolearner/hypertorch/blob/ceb4b0964df6b7ad87b680b267b671d166837b30/examples/iMAML.py#L164-L166)? In particular I am not sure what approximation of the paper you are referring to?
- You use the batch norm after the activation and the pooling, unlike the original implementation, why?
- Again a question regarding the difference between your implementation and the original: you initialize linear weights at zero, why?

Thanks so much for taking the time to read me it's very much appreciated.

Cheers!

**Comment by prolearner** [comment]
Hi @zaccharieramzi,
thanks a lot for your interest in the code :).

I did not try to reproduce the exact numbers in the iMAML paper, but just to implement the main components of the experimental setup. The purpose of the iMAML example is just to showcase what it can be achieved with the hypergradient approximation methods implemented here and possibly be used as a starting point for further research or to apply the method to different data. 

As you noticed there are some differences in the implementation which happened mostly by chance. You can try to modify the implementation to match the one described by the original paper and code to see if this improves accuracy. I'd try first with larger batch size and using the same network (with correct placement of batch norm), possibly tweaking the learning rate. I don't think initialising the final linear layer weights randomly instead of to zero would make much difference.

The approximation I'm referring to [here](https://github.com/prolearner/hypertorch/blob/ceb4b0964df6b7ad87b680b267b671d166837b30/examples/iMAML.py#L164-L166) is the hypergradient approximation that they use, which relies on  the conjugate gradient method to solve the linear system.

**Comment by zaccharieramzi** [comment]
Ok makes total sense!
Problem is that even with the original implementation I cannot retrieve the numbers stated in the paper.

Will let you know if I get to anything with my own implementation.
